x-pack/platform/packages/shared/kbn-streamlang/src/validation/validations.ts

```diff
@@ -0,0 +1,260 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { noop } from 'lodash';
+import type { Condition } from '../../types/conditions';
+import {
+  isAndCondition,
+  isOrCondition,
+  isNotCondition,
+  isFilterCondition,
+  isAlwaysCondition,
+} from '../../types/conditions';
+import type {
+  StreamlangProcessorDefinition,
+  ProcessorType,
+  AppendProcessor,
+  ConvertProcessor,
+  DateProcessor,
+  DissectProcessor,
+  GrokProcessor,
+  RenameProcessor,
+  SetProcessor,
+  RemoveProcessor,
+  RemoveByPrefixProcessor,
+  ReplaceProcessor,
+} from '../../types/processors';
+import type { StreamlangStep } from '../../types/streamlang';
+import {
+  isActionBlock,
+  isWhereBlock,
+  streamlangDSLSchema,
+  type StreamlangDSL,
+  type StreamType,
+} from '../../types/streamlang';
+
+export class StreamlangValidationError extends Error {
+  constructor(public readonly errors: string[]) {
+    super(errors.join('\n'));
+    this.name = 'StreamlangValidationError';
+  }
+}
+
+/**
+ * Format Zod errors into human-readable error messages.
+ */
+function formatZodErrors(zodError: {
+  issues: Array<{ path: (string | number)[]; message: string }>;
+}): string[] {
+  return zodError.issues.map((issue) => {
+    const path = issue.path.length > 0 ? issue.path.join('.') : 'root';
+    return `${path}: ${issue.message}`;
+  });
+}
+
+export const validateStreamlang = (dsl: StreamlangDSL, streamType?: StreamType): true => {
```

Comment:
Your `validateStreamlang` is pretty close to mine in terms of how it's integrated, they should fit right together.